Generic event collector (#37806)
demisto/content: Packs/GenericAPIEventCollector/Integrations/GenericAPIEventCollector/README.md

## Configure Generic API Event Collector (Beta) in Cortex

| **Parameter**                                                              | **Description**                                                                                                                                                                                                                                                                            | **Required** |
|----------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------|
| Server URL                                                                 |                                                                                                                                                                                                                                                                                            | True         |

> | Endpoint                                                                   | Add the endpoint you want to collect data from \(Alert/Events Etc..\)                                                                                                                                                                                                                      | True         |

Reviewer: ```suggestion
| Endpoint                                                                   | Add the endpoint you want to collect data from \(Alert/Events etc.\).                                                                                                                                                                                                                      | True         |
```

> | Authentication Type                                                        | Select the authentication method                                                                                                                                                                                                                                                           | True         |

Reviewer: ```suggestion
| Authentication Type                                                        | Select the authentication method.                                                                                                                                                                                                                                                           | True         |
```

| API Token                                                                  | API Key to access the service REST API.                                                                                                                                                                                                                                                    | False        |
| Username                                                                   | Username &amp;amp; Password to use for basic authentication.                                                                                                                                                                                                                               | False        |
| Password                                                                   |                                                                                                                                                                                                                                                                                            | False        |

> | Add Fields To header                                                       | If the product authentication requires more fields to add to the header please add it here in a dictionary format \{unique_field : 286\} if there's a need to add more then one use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\}                | False        |

Reviewer: ```suggestion
| Add Fields To header                                                       | If the product authentication requires more fields to add to the header add it here in dictionary format \{unique_field : 286\}. If there's a need to add more then one, use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\}                | False        |
```

| Use system proxy settings                                                  |                                                                                                                                                                                                                                                                                            | False        |
| Trust any certificate (not secure)                                         |                                                                                                                                                                                                                                                                                            | False        |

> | Vendor                                                                     | Enter vendor name for dataset                                                                                                                                                                                                                                                              | True         |
> | Product                                                                    | Enter product name for dataset                                                                                                                                                                                                                                                             | True         |

Reviewer: ```suggestion
| Vendor                                                                     | Enter vendor name for dataset.                                                                                                                                                                                                                                                              | True         |
| Product                                                                    | Enter product name for dataset.                                                                                                                                                                                                                                                            | True         |
```

> | Request data                                                               | If the product authentication requires more fields to add to the DATA please add it here in a dictionary format \{unique_field : 286\} if there's a need to add more then one use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\}                  | False        |

Reviewer: ```suggestion
| Request data                                                               | If the product authentication requires more fields to add to the DATA add it here in dictionary format \{unique_field : 286\}. If there's a need to add more then one, use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\}                  | False        |
```

> | Initial request data                                                       | If the product requires a different initial DATA,  please add it here in a dictionary format \{unique_field : 286\} if there's a need to add more then one use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\}                                     | False        |

Reviewer: ```suggestion
| Initial request data                                                       | If the product requires a different initial DATA,  add it here in dictionary format \{unique_field : 286\}. If there's a need to add more then one, use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\}                                     | False        |
```

> | Request JSON parameters                                                    | If the product authentication requires more fields to add to the body as JSON please add it here in a dictionary format \{unique_field : 286\} if there's a need to add more then one use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\}          | False        |
> | Initial request JSON parameters                                            | If the product requires a different initial request JSON, please add it here in a dictionary format \{unique_field : 286\} if there's a need to add more then one use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\}                              | False        |
> | Query parameters                                                           | If the product authentication allows to filter the results using query Parameters please add it here in a dictionary format \{unique_field : 286\} if there's a need to add more then one use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\}      | False        |
> | Initial query parameters                                                   | If the product requires a different initial query parameters for the first fetch call, please add it here in a dictionary format \{unique_field : 286\} if there's a need to add more then one use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\} | False        |

Reviewer: ```suggestion
| Request JSON parameters                                                    | If the product authentication requires more fields to add to the body as JSON add it here in dictionary format \{unique_field : 286\}. If there's a need to add more then one, use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\}          | False        |
| Initial request JSON parameters                                            | If the product requires a different initial request JSON, add it here in dictionary format \{unique_field : 286\}. If there's a need to add more then one, use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\}                              | False        |
| Query parameters                                                           | If the product authentication allows to filter the results using query Parameters add it here in dictionary format \{unique_field : 286\}. If there's a need to add more then one, use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\}      | False        |
| Initial query parameters                                                   | If the product requires a different initial query parameters for the first fetch call, add it here in dictionary format \{unique_field : 286\}. If there's a need to add more then one, use it in this format: \{'field-1': value_example, 'field-2': value_2, 'field-3': value_3\} | False        |
```

| Is pagination needed                                                       | If the API JSON response supports events pagination.                                                                                                                                                                                                                                       |              |
| Pagination field name                                                      | Next page field in JSON response, e.g., "cursor", "next_page"                                                                                                                                                                                                                              | False        |
| Pagination flag                                                            | Next page existence in JSON response e.g., "has_more", "next"                                                                                                                                                                                                                              | False        |

> | Timestamp format of the event creation time or "epoch".                    | Python compatible datetime formatting \(e.g. "%Y-%m-%dT%H:%M:%S.%fZ" or "%Y.%m.%d %H:%M:%S"\) or "epoch" to use UNIX epoch time.                                                                                                                                                           | False        |

Reviewer: ```suggestion
| Timestamp format of the event creation time or "epoch".                    | Python compatible datetime formatting \(e.g. ,"%Y-%m-%dT%H:%M:%S.%fZ" or "%Y.%m.%d %H:%M:%S"\) or "epoch" to use UNIX epoch time.                                                                                                                                                           | False        |
```

> | Timestamp field                                                            | The name of the event creation time in the response data, e.g. "timestamp" or "created_at".                                                                                                                                                                                                | True         |

Reviewer: ```suggestion
| Timestamp field                                                            | The name of the event creation time in the response data, e.g., "timestamp" or "created_at".                                                                                                                                                                                                | True         |
```

> | Events lookup path in the response JSON, dot seperated, e.g. "data.items". | Where within the response object to find the events list.                                                                                                                                                                                                                                  | False        |

Reviewer: ```suggestion
| Events lookup path in the response JSON, dot-separated, e.g., "data.items". | Where within the response object to find the events list.                                                                                                                                                                                                                                  | False        |
```

> | Event ID lookup path in the event response JSON, dot seperated, e.g. "id". | Where within the event object to find the events id.                                                                                                                                                                                                                                       | False        |
> | The type of ID field, either "integer" or "string"                         | ID field of type integer are comparable and when last fetched id is the maximum ID between the fetched events, when the type is string, the last fetched ID is the last event returned from the API.                                                                                       | False        |

Reviewer: ```suggestion
| Event ID lookup path in the event response JSON, dot-separated, e.g., "id". | Where within the event object to find the event ID.                                                                                                                                                                                                                                       | False        |
| The type of ID field, either "integer" or "string"                         | ID field of type integer are comparable and when last fetched ID is the maximum ID between the fetched events, when the type is string, the last fetched ID is the last event returned from the API.                                                                                       | False        |
```

| Limit                                                                      | Number of incidents to fetch per fetch.                                                                                                                                                                                                                                                    | False        |
| Fetch Events                                                               |                                                                                                                                                                                                                                                                                            | False        |
| Events Fetch Interval                                                      |                                                                                                                                                                                                                                                                                            | False        |

> How to configure the event collector

Reviewer: ```suggestion
## How to configure the event collector
```

---

### Authentication
You must specify the authentication method required by the server. 
The supported authentication types include:

> 1. Basic authentication (username and password)
> 2. Token-based authentication
> 3. Bearer token
> 4. Api-Key token
> 5. Raw Token (for custom token-based authentication)
> 6. No Authorization (for publicly accessible data)

Reviewer: ```suggestion
- Basic authentication (username and password)
- Token-based authentication
- Bearer token
- Api-Key token
- Raw Token (for custom token-based authentication)
- No Authorization (for publicly accessible data)
```

### Pagination

> When the API supports pagination in the response, the collector can fetch more pages of data using the below parameters

Reviewer: ```suggestion
When the API supports pagination in the response, the collector can fetch more pages of data using the following parameters:
```

> 1. Is pagination needed If the API JSON response supports events pagination.
> 2. Pagination field name, Next page field in JSON response, e.g., "cursor" or "next_page" | False |
> 3. Pagination flag, The Next page existence in JSON response e.g., "has_more" or "next"

Reviewer: ```suggestion
- Is pagination needed If the API JSON response supports events pagination.
- Pagination field name, Next page field in JSON response, e.g., "cursor" or "next_page" | False |
- Pagination flag, The Next page existence in JSON response e.g., "has_more" or "next"
```

In the below example the pagination flag is `pagination.has_more`
The pagination field name is `pagination.next_page`
```json
{
    "data": [
        {
            "id": 1,
            "name": "John Doe",
            "occupation": "Software Engineer"
        },
        {
            "id": 2,
            "name": "Jane Smith",
            "occupation": "Data Scientist"
        }
    ],
    "pagination": {
        "current_page": 1,
        "next_page": "https://api.example.com/users?page=2",
        "has_more": true
    }
}
```

### Request Data (And initial request data)
If the product authentication requires more fields to add to the `DATA`.

> Please add it here in a dictionary format.

Reviewer: ```suggestion
Add it here in dictionary format.
```

For example:
```json
{"field-1": "value_example", "field-2": 1, "field-3": "value_3"}
```
Note: Using the initial request data parameter will only be used in the first request to collect events.
### Request JSON (And initial request JSON)
If the product authentication requires more fields to add to the body as JSON, please add it 

> here in a dictionary format.

Reviewer: ```suggestion
here in dictionary format.
```

For example:
```json
{"date": "2021-08-01", "field-2": 1, "field-3": "value_3"}
```

Note: Using the initial request JSON parameter will only be used in the first request to collect events.

### Query parameters (And Initial Query parameters)

> If the product authentication allows filtering the results using query parameters, please add it here in a dictionary format:

Reviewer: ```suggestion
If the product authentication allows filtering the results using query parameters, add it here in dictionary format:
```

```json
{"ordering": "id", "limit": 1, "created_after": "@first_fetch_datetime"}
```

Note: Using the initial query parameters parameter will only be used in the first request to collect events.

### Timestamp field

> The name of the event creation time in the response data, e.g. "timestamp" or "created_at".

Reviewer: ```suggestion
The name of the event creation time in the response data, e.g., "timestamp" or "created_at".
```

In the following API response:
```json
{
  "data": [
    {
      "id": 3,
      "name": "Alice Brown",
      "occupation": "Network Engineer",
      "created": "2021-10-05T19:45:20.789012Z"
    },
    {
      "id": 4,
      "name": "Dave Wilson",
      "occupation": "Cybersecurity Analyst",
      "created": "2021-10-06T10:15:45.654321Z"
    }
  ],
  "pagination": {
    "current_page": 2,
    "next_page": "https://api.example.com/users?page=3",
    "has_more": true
  }
}
```
the timestamp field is `created`

### Timestamp format
The timestamp format of the event creation time or "epoch" to use UNIX epoch time.

> The formatting supported is Python-compatible datetime formatting (e.g. "%Y-%m-%dT%H:%M:%S.%fZ" or "%Y.%m.%d %H:%M:%S").

Reviewer: ```suggestion
The formatting supported is Python-compatible datetime formatting (e.g., "%Y-%m-%dT%H:%M:%S.%fZ" or "%Y.%m.%d %H:%M:%S").
```

In the following API response:
```json
{
  "data": [
    {
      "id": 3,
      "name": "Alice Brown",
      "occupation": "Network Engineer",
      "created": "2021-10-05T19:45:20.789012Z"
    },
    {
      "id": 4,
      "name": "Dave Wilson",
      "occupation": "Cybersecurity Analyst",
      "created": "2021-10-06T10:15:45.654321Z"
    }
  ],
  "pagination": {
    "current_page": 2,
    "next_page": "https://api.example.com/users?page=3",
    "has_more": true
  }
}
```
The timestamp format is python format "%Y-%m-%dT%H:%M:%S.%fZ"

Note: To learn more about Python date and time formats, see: https://docs.python.org/3/library/datetime.html#format-codes 

### Events

> Where within the response JSON to search for the events, dot seperated (e.g. "data.items").

Reviewer: ```suggestion
Where within the response JSON to search for the events, dot-separated (e.g., "data.items").
```

Example 1:
```json
{
  "data": [
    {
      "id": 4,
      "name": "Alice Brown",
      "occupation": "Network Engineer",
      "created": "2021-10-05T19:45:20.789012Z"
    },
    {
      "id": 3,
      "name": "Dave Wilson",
      "occupation": "Cybersecurity Analyst",
      "created": "2021-10-06T10:15:45.654321Z"
    }
  ]
}
```
The events are within the "data" in the response.

> Example 1:

Reviewer: ```suggestion
Example 2:
```

```json
{
  "data": {
    "items": [
        {
            "id": 4,
            "name": "Alice Brown",
            "occupation": "Network Engineer",
            "created": "2021-10-05T19:45:20.789012Z"
        },
        {
            "id": 3,
            "name": "Dave Wilson",
            "occupation": "Cybersecurity Analyst",
            "created": "2021-10-06T10:15:45.654321Z"
        }
    ]
  }
}
```
The events are within the "data.items" in the response.

### Event ID & Type

> Event ID lookup path in the event response JSON, dot seperated, e.g. "id"
> Where within the event object to find the events' id.

Reviewer: ```suggestion
Event ID lookup path in the event response JSON, dot-separated, e.g., "id"
Where within the event object to find the event ID.
```

The type of ID field, either "integer" or "string":

> - ID field of type integer is comparable, and when last fetched id is the maximum ID between the fetched events.

Reviewer: ```suggestion
- ID field of type integer is comparable, and when last fetched ID is the maximum ID between the fetched events.
```

- ID field of the type is string, the last fetched ID is the last event returned from the API.
  
Example 1:
```json
{
  "data": [
    {
      "id": 4,
      "name": "Alice Brown",
      "occupation": "Network Engineer",
      "created": "2021-10-05T19:45:20.789012Z"
    },
    {
      "id": 3,
      "name": "Dave Wilson",
      "occupation": "Cybersecurity Analyst",
      "created": "2021-10-06T10:15:45.654321Z"
    }
  ],
  "pagination": {
    "current_page": 2,
    "next_page": "https://api.example.com/users?page=3",
    "has_more": true
  }
}
```

> The event id field should be "id" and the type should be integer, and the last fetched id will be 4.

Reviewer: ```suggestion
The event ID field should be "id" and the type should be integer, and the last fetched ID will be 4.
```

Example 2:
```json
{
  "data": [
    {
      "uuid": "123e4567-e89b-12d3-a456-426614174000",
      "name": "Alice Brown",
      "occupation": "Network Engineer",
      "created": "2021-10-05T19:45:20.789012Z"
    },
    {
      "uuid": "123e4567-e89b-12d3-a456-426614174001",
      "name": "Dave Wilson",
      "occupation": "Cybersecurity Analyst",
      "created": "2021-10-06T10:15:45.654321Z"
    }
  ],
  "pagination": {
    "current_page": 2,
    "next_page": "https://api.example.com/users?page=3",
    "has_more": true
  }
}
```

> The event id field should be "uuid" and the type should be string, and the last fetched id will be "123e4567-e89b-12d3-a456-426614174001".

Reviewer: ```suggestion
The event ID field should be "uuid" and the type should be string, and the last fetched ID will be "123e4567-e89b-12d3-a456-426614174001".
```

## Substitutions in API requests calls
To make the API calls more dynamic against the API endpoint, we added a few placeholders that will be substituted before calling the API endpoint.

> 1. `@last_fetched_id` - The last id that was fetched from the API, if this is the first fetch, the value will be empty.
> 2. `@last_fetched_datetime` - The last fetched event time from the API, if this is the first fetch, the value will be empty.
> 3. `@first_fetch_datetime` - The first fetch time, when the integration first started to fetch events.
> 4. `@fetch_size_limit` - The number of incidents to fetch per fetch.

Reviewer: ```suggestion
- `@last_fetched_id` - The last ID that was fetched from the API, if this is the first fetch, the value will be empty.
- `@last_fetched_datetime` - The last fetched event time from the API, if this is the first fetch, the value will be empty.
- `@first_fetch_datetime` - The first fetch time, when the integration first started to fetch events.
- `@fetch_size_limit` - The number of incidents to fetch per fetch.
```

Examples being used in query parameters:

> - This will substitute the `@last_fetched_id` with the last fetched id from a previous fetch call.

Reviewer: ```suggestion
- This will substitute the `@last_fetched_id` with the last fetched ID from a previous fetch call.
```
